jfrog-cli/artifactory/commands/docker/push.go

```diff
@@ -65,3 +76,39 @@ func createServiceManager(config *DockerPushConfig) (*artifactory.ArtifactorySer
 
 	return artifactory.New(serviceConfig)
 }
+
+type dockerLoginConfig struct {
+	ArtifactoryDetails *config.ArtifactoryDetails
+}
+
+// First will try to login assuming a proxy-less tag (e.g. "registry-address/docker-repo/image:ver").
+// If fails, we will try assuming a reverse proxy tag (e.g. "registry-address-docker-repo/image:ver").
+func dockerLogin(imageTag string, config *dockerLoginConfig) error {
+	imageRegistry, err := docker.ResolveRegistryFromTag(imageTag)
+	if err != nil {
+		return err
+	}
+
+	cmd := &docker.LoginCmd{DockerRegistry: imageRegistry, Username: config.ArtifactoryDetails.User, Password: config.ArtifactoryDetails.Password}
+	err = utils.RunCmd(cmd)
+
+	if exitCode := cliutils.GetExitCode(err, 0, 0, false); exitCode == cliutils.ExitCodeNoError {
+		// Login succeeded
+		return nil
+	}
+
+	indexOfSlash := strings.Index(imageTag, "/")
+	if indexOfSlash < 0 {
+		return errorutils.CheckError(errors.New("Docker login failed for provided tag."))
```

Comment:
I suggest we go for a less cryptic error message here.
Let's come up with a good message which explains the expected docker tag format and covers both proxy-less and reverse-proxy options.